### PR TITLE
Fixed: remove HTTP2 pseudo-headers when propagating HTTP/2 request to HTTP/1 service

### PR DESCRIPTION
**Pull request by Dogild**
- [x] Description and context for reviewers: one partner, one stranger

In case grpc inbound (HTTP/2) ->  http outbound (HTTP/1) (for instance with a YARPC middleware), and the request contains http2 specific pseudo-header fields, the http outbound call will fail. 

All pseudo-headers are being removed for the request.

The added test will fail before the fix:

```
--- FAIL: TestGRPCInHTTPOut (0.00s)
    outbound_test.go:676:
        	Error Trace:	outbound_test.go:676
        	Error:      	Received unexpected error:
        	            	code:unknown message:unknown error from http client: Post "http://127.0.0.1:63722": net/http: invalid header field name "Rpc-Header-:path"
        	Test:       	TestGRPCInHTTPOut
```

It should be noted that this fix will be harmless for services. net/http rejects already header with rune `:`. With this diff, in the worst case we let more request to go.

- [x] Entry in CHANGELOG.md

Note: this is a duplicate of https://github.com/yarpc/yarpc-go/pull/2026 with refactoring, rebase and more comments

### CODE REVIEW

File: transport/http/outbound_test.go [hunk 1]
```diff
@@ -683,6 +683,47 @@ func TestServiceMatchSuccess(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestGRPCInHTTPOut(t *testing.T) {
```
Comment (prashantv) [inline transport/http/outbound_test.go:686]: (follow-up integration test idea) Should we have an integration test which verifies a gRPC request being forwarded to a HTTP outbound (and probably should do that for all combinations actually). Probably doesn't belong in transport/http but at the top-level probably

**Review by prashantv — APPROVED** [review]
Since the deleted headers are overridden by the YARPC HTTP outbound, deleting the headers seems OK.

Only question I have is whether these should be deleted in the gRPC inbound (though that has a larger impact if users are relying on them).

**Comment by Dogild** [comment]
> Since the deleted headers are overridden by the YARPC HTTP outbound, deleting the headers seems OK.
> 
> Only question I have is whether these should be deleted in the gRPC inbound (though that has a larger impact if users are relying on them).

Yeh we would hit a regression if users are consuming these headers (either in middleware or handler).

File: transport/http/outbound_test.go (hunk 1 above)
Comment (Dogild) [reply transport/http/outbound_test.go:686]: Agree with this, I will take care of this in another PR though (less simple to review). The idea would be to have tests where request are being forwarded in a transport middleware where outbound transport will be different. 

This test actually checks how headers are actually updated or removed. I will update the name of the test though.

**Pull request merged**